### ISSUE TITLE
Add applicable Vitess version to manual pages / documentation

### ISSUE THREAD

**Issue by bolt-leandro** [vitessio/website#625]
#### Feature Description

Vitess is currently changing at a fast pace so some parts of the written documentation are not backwards compatible or only apply to specific versions of Vitess.
It would be good if when writing/updating documentation, author could attach the Vitess version it applies to. A simple example is the MySQL docs where there is version selector ([https://dev.mysql.com/doc/refman/8.0/en/](https://dev.mysql.com/doc/refman/8.0/en/)). 

#### Use Case(s)

For example, currently, Vitess 8.0 gh-ost integration has specialized syntax while newer versions have @@ddl_strategy. The 8.0 GA version and the 'Managed, Online Schema Changes' man page are out of sync. One would have to check man page version history to find out about deprecated specialized syntax . This would probably cause some confusion to unaware users.

**Comment by deepthi** [comment]
I looked at kubernetes.io and it seems like there may be a standard way of doing docs versioning with Hugo.
@doeg any thoughts on this?

**Comment by doeg** [comment]
It looks like there are (at least?) two approaches to versioned content with Hugo. My $0.02 is that approach the first (git tags + separate deploys) is _probably_ simpler to set up, though perhaps less elegant from a user's perspective. Ultimately I think it'd be worth timeboxing an investigation into both to see which works best in practice. I could see the devil being in the details, particularly in terms of search + navigation. 

**1. Maintain previous versions as git tags + deploy static shapshots.**

I _think_ this is what https://kubernetes.io is doing. Specifically:

- [Docsy](https://www.docsy.dev/docs/adding-content/versioning/) is used for the version dropdown in the navbar
- Older versions of kubernetes.io documentation are deployed as a "static snapshot", for example: https://v1-15.docs.kubernetes.io.

One thing to note here is that the snapshots will be deployed with whatever layout they had at the time. 

**2. Version documentation with folder structure.**

Another approach is to "archive" docs using folder structure. For example:

content/docs/
├─ latest
├─ v8.0
├─ v7.0

This write-up summarizes some of the nuances with this approach, especially with links: https://github.com/google/docsy/issues/114#issuecomment-516592637

This thread is also useful + mentions how versioning works with the search widget: https://discourse.gohugo.io/t/documentation-site-versioning/5898/7

**Comment by doeg** [comment]
I like the way Cordova does it: https://cordova.apache.org/docs/en/9.x/ It looks like they use Jekyll, though. 🤔 https://github.com/apache/cordova-docs

<img width="2672" alt="Screen Shot 2020-12-03 at 9 00 16 PM" src="https://user-images.githubusercontent.com/855595/101112094-924a7e00-35aa-11eb-98f2-acc76070a61a.png">

**Comment by hallaroo** [comment]
Band aid measure to update all the PDFs of the older vitess versions is here https://github.com/vitessio/website/pull/632

**Comment by deepthi** [comment]
@doeg thank you for looking into this. It will be great if you can investigate further. I'm available to discuss / bounce ideas off of.

**Comment by shlomi-noach** [comment]
Hello :wave: I have some thoughts on this topic and wish to take it on myself to push this through (thanks @deepthi for pointing me to this discussion).

There are several topics here to consider:

1. First and foremost, how do the development team / the maintainers even know which version each doc should apply to
2. What's an easy process for the developers/maintainers to assign docs to versions, in practice?
3. How do we present versioned docs to the user?
4. What changes do we need to make to our website platform to support that?

I have ideas that will simplify the process for the developers, without creating much hassle on the website/framework side.

## Versioned docs

I believe the source of truth for the Vitess docs should be in the vitess repo. This, IMO, elegantly simplifies documentation flow:

- When a developer makes changes to the code, they make matching changes in documentation, on the very same Pull Request. (today, with separate repos, we need to cross reference PRs, validate that a code PR has a matching doc PR, remember to make doc changes...)
- If we tag, say, `release-9.0`, then both code and documentation are tagged together. Since we the developers use that flow _anyhow_ for our code, we get documentation versioning for free.
- If we backport code into an earlier version, we also backport documentation. Again, this comes free.

## User experience

I agree with @bolt-leandro that the MySQL documentation provides a simple and effective interface to browse through versions:

![Screenshot from 2020-12-21 12-33-28](https://user-images.githubusercontent.com/2607934/102767796-cda9c280-4388-11eb-8245-22cd85367dee.png)

![Screenshot from 2020-12-21 12-33-14](https://user-images.githubusercontent.com/2607934/102767806-d1d5e000-4388-11eb-94c0-93d22d6ae673.png)

## Minimal changes to the framework

In the above screenshots you can see that the version translates to a URL path, which is compatible with @doeg 's "Version documentation with folder structure." suggestion.

Moreover, _we already use this technique_ with website languages. Consider https://vitess.io/docs/get-started/local/ vs https://vitess.io/zh/docs/get-started/local/. We already maintain a different path for a different language; we could do the same for a different version.

# Implementations

And this would be mostly automated. My idea is to:

- keep this repo (https://github.com/vitessio/website) as is. 
- The source of truth for Vitess docs will be on https://github.com/vitessio/vitess/
- for every merge to `master` branch, we will run an automated `push` from https://github.com/vitessio/vitess/ (`master`) to https://github.com/vitessio/website (`prod`)
- as well as any `release-*` content to a compatible path in https://github.com/vitessio/website, again on `prod`

This will effectively keep docs in sync at all times.

Noteworthy:

- https://github.com/vitessio/vitess will only contain the vitess docs, not the entire website
- changes to the docs will _have to be made_ on the Vitess repo, https://github.com/vitessio/vitess
- changes to non-docs, e.g. blogs, general website content, design etc. will be made to https://github.com/vitessio/website

**Comment by shlomi-noach** [comment]
Hugo doesn't have support for this kind of multi-version site; what I've [seen](https://discourse.gohugo.io/t/develop-and-generate-different-site-versions-from-a-single-site/14628/15) is people piggy riding the languages feature; so `8.0` will be a "language". This actually falls in line with how MySQL does that, so I'm gonna give it a try.

**Comment by shlomi-noach** [comment]
Hugo is fighting against me. I'm trying to add a "language" called `Current` and a language called `8.0`, that's fine. But can't make `Current` the default language NoMatterWhat™; `DefaultContentLanguage = "current"` seems to have no effect, the site always defaults to English, even if I remove `English` altogether.

**Comment by shlomi-noach** [comment]
Just to point out some issues I see with the implementation:

1. For convenience, documentation developers will want to still use Vercel's preview for PRs. This means that for any `vitessio/vitess` PR, we will need to open a correlated `website` PR.
  - this could be done by automation (complexity: identify that a PR already exists; complexity: identify that there's no need for a PR if docs are unchanged in the `vitess` PR, or if docs changes have been removed)
  - or we can at first create a script in the `vitess` repo that the developer will execute, and which will open the PR on their behalf.
2. we must prevent merging of such PRs on `vitessio/website`
3. There may be direct PRs against non docs parts of `vitess/website` (e.g. `blog`)
4. but we must reject PRs that incorporate changes to both `docs` and non-`docs`
5. probably more to come as I investigate this.

I think I'll begin with a custom script that the developer will run. It's a soft entry point before going fully-automated.

**Comment by nate-double-u** [comment]
Hi, I'm Nate w/ the CNCF TechDocs team 👋. @deepthi [asked us to take a look at this issue](https://cncfservicedesk.atlassian.net/browse/CNCFSD-947).

## tl;dr

I'd suggest using a folder based approach to versioning, and keeping the primary, canonical, documentation in this website repo, separate from the code repo. 

As this site uses Hugo, but not Docsy, I think that the [cni.dev](https://www.cni.dev/plugins/current/) docs site is a good example of how versioning could work here, though some changes would have to be made to accommodate the multiple languages supported here.

## Detailed responses and suggestions

I've done some [research on how best to do versioning on hugo sites](https://github.com/cncf/techdocs/blob/main/howto/versioning-documentation.md), and there's mostly two ways to handle it with hugo (as @doeg [mentions above](https://github.com/vitessio/website/issues/625#issuecomment-738504396)), folder based, and sub-domain based. 

Hugo defaults to sub-domain based, and this is what Kubernete's website uses -- one of the main complications of this method is that it requires the website repository to maintain feature branches for all the versions that are supported. These feature branches would each need to be updated with any structural changes to the site (for instance any changes to the menu or style would need to be done to each previous version's feature branch). It also versions the _whole_ site, not just the docs, which may not be ideal.

I've set up several sites that use folder based versioning, a couple examples: [etcd.io](https://etcd.io) (which uses the docsy theme), and [cni.dev](https://cni.dev), which started with subdomain versioning, then was converted to folder based.

I tend to recommend folder based versioning as it's easier to setup and maintain, though it does make search a little more complex.

@shlomi-noach asks some good questions:

> There are several topics here to consider:
>    1. First and foremost, how do the development team / the maintainers even know which version each doc should apply to

I'd recommend having a docs version for every minor release (following semver's MAJOR.MINOR.PATCH). Things that need to be made based on a patch release of the code can just be added to the minor version in code.

>    2. What's an easy process for the developers/maintainers to assign docs to versions, in practice?

If you run a folder based versioning system, it also makes supporting Current +X number of versions a little easier, as your current is just Major.Minor -- then when it comes time to release the next Minor release, you can copy the folder over as a starting point. All updates go into the latest major.minor version of the docs until it's time for a new release -- at this point it's probably a good idea to add documentation updates to the main release process.

It looks like you've already got PDF archives of the docs for every Major release, which may be a good cadence for this too.

>    3. How do we present versioned docs to the user?

I think https://etcd.io/docs/v3.5/ does this well - with the versions on the side bar, and a version selector dropdown up top. Again this is partly Docsy, but it may be able to be implemented here w/o changing the theme.

>    4. What changes do we need to make to our website platform to support that?

When you suggest that "I believe the source of truth for the Vitess docs should be in the vitess repo.", do you mean the [vitessio/vitess](https://github.com/vitessio/vitess) repo? If so, I'd suggest against that (unless you want that repo to become a mono-repo of sorts, where the website would be a subfolder of it and get deployed from that repo directly). I suggest keeping them separate and having this [vitessio/website](https://github.com/vitessio/website/) repo stay the source  of truth for the documentation. The .md files that reside in this folder really are designed as website pages, and having them exist in another repo tends towards confusion -- links don't work right, etc. The other reason to keep it separate is localization. The site currently supports 2 languages, and can support more, but if the docs are moved into the same repo as the code, we potentially lose the ability to easily switch between languages.

**Comment by shlomi-noach** [comment]
Thanks @nate-double-u for your thoughts!

> If so, I'd suggest against that (unless you want that repo to become a mono-repo of sorts, where the website would be a subfolder of it and get deployed from that repo directly)

Yes, that was exactly what I was thinking. Having the website/docs as a subfolder in https://github.com/vitessio/vitess and generating the website from there. As an alternative, keep this (https://github.com/vitessio/website) repo up and do a nightly sync from https://github.com/vitessio/vitess to this repo, then generate the webstie from here.

We recently did a small change to the release notes flow in https://github.com/vitessio/vitess, which is in my opinion a small step towards my objective: any PR can be labeled with "release notes - needs notes" or similar. A PR that is labeled as such, is then required to edit the release notes (which are inside https://github.com/vitessio/vitess) to reflect the changes. I think it works very well for us, and it makes me document, in a very small way, my changes in the same PR they're submitted in.

IMO the documentation should likewise be coupled with the code.

I agree that many files in _this_ repo are irrelevant to https://github.com/vitessio/vitess, such as the hugo templates etc. But I think `.md` files for the docs themselves still make perfect sense and are today's de-facto documentation format on GitHub.

Last, my very personal opinion on multi-languages (I don't necessarily reflect anyone else's opinion here): I don't think it works. I think we should stick with English. Today, we don't have authoritative people who can do documentation in Chinese, for example. The Chinese docs are always out-of-date, with casual sprints by external contributors (much appreciated!) to catch up. When that happens, I still don't think _everything_ gets caught up. And so for me and for the team, there's no way we can tell that the docs are good. Which leads me to believe we should stop supporting extra languages, at least until we reach a point of maturity where we are able to have contributors who do that as their day job.

**Comment by nate-double-u** [comment]
> Thanks @nate-double-u for your thoughts!

Happy to chip in 🙂

> > If so, I'd suggest against that (unless you want that repo to become a mono-repo of sorts, where the website would be a subfolder of it and get deployed from that repo directly)
> 
> Yes, that was exactly what I was thinking. Having the website/docs as a subfolder in https://github.com/vitessio/vitess and generating the website from there. As an alternative, keep this (https://github.com/vitessio/website) repo up and do a nightly sync from https://github.com/vitessio/vitess to this repo, then generate the webstie from here.
> 
> ...
> 
> IMO the documentation should likewise be coupled with the code.

I think this is probably outside the scope of this issue. I think that regardless of where the site lives, we can make versioning work. Maybe we should open a separate issue to discuss moving the documentation with the community?

> 
> Last, my very personal opinion on multi-languages (I don't necessarily reflect anyone else's opinion here): I don't think it works. I think we should stick with English. Today, we don't have authoritative people who can do documentation in Chinese, for example. The Chinese docs are always out-of-date, with casual sprints by external contributors (much appreciated!) to catch up. When that happens, I still don't think _everything_ gets caught up. And so for me and for the team, there's no way we can tell that the docs are good. Which leads me to believe we should stop supporting extra languages, at least until we reach a point of maturity where we are able to have contributors who do that as their day job.

I'm new to this project, but I here are my general thoughts on localization:

I think it would be a shame to throw away all that work. I think we can make versioning work with multiple languages. It was a different site, but [I was able to get versioning working with localization on a docsy example site](https://github.com/nate-double-u/docsy-example/pull/1) ([deploy preview](https://deploy-preview-1--jolly-albattani-baa029.netlify.app/docs/current/overview/)), it shouldn't be too different here.

Localization always lags behind a docs set's primary language. That is another argument for folder based versioning, each language group can translate versions as they're able without holding back another language group.

As for needing professionals to do localization and translation, I'm not sure that's needed -- the primary docs aren't written fully by pros, and the language community that is using the docs are able to open issues against incorrect docs themselves so I'm not sure that they would be more or less incorrect than the primary language. 

Analytics may help here too, see if the localizations are well used and worth the maintenance cost.

This is another good point to suggest opening an issue about. Kubernetes has [a process for localization](https://kubernetes.io/docs/contribute/localization/), and a documentation subgroup to help manage it. If there is community interest, rather than throwing the translation away, it may be better to formalize it.

**Comment by deepthi** [comment]
> I think this is probably outside the scope of this issue. I think that regardless of where the site lives, we can make versioning work. Maybe we should open a separate issue to discuss moving the documentation with the community?

+1. It is worth getting the versioning working by itself regardless of the original source of the docs.

**Comment by shlomi-noach** [comment]
> I think this is probably outside the scope of this issue. I think that regardless of where the site lives, we can make versioning work. Maybe we should open a separate issue to discuss moving the documentation with the community?

Agreed.

> Analytics may help here too, see if the localizations are well used and worth the maintenance cost.

That's a great suggestion!

**Comment by nate-double-u** [comment]
I have some time & can work on this -- so my goals would be:

- [ ] Setup folder based versioning
  - [ ] Only docs (blog, community, and other pages left un-versioned) 
  - [ ] Current +1 (with instructions on how to add more)
  - [ ] Respects the current locales [en & zh]

**Comment by doeg** [comment]
Thank you for working on this and for your thorough notes, Nate! This looks great + I'm really excited to see it happen.
